### PR TITLE
Add basics of legacy XML coverpage generation using .legacyxml format

### PR DESCRIPTION
**Pull request by longhotsummer**
You'll need to add a published document to Kenya for this to work. The go to `/api/v1/ke/act/xxx/yyy.legacyxml` for whatever xxx and yyy are for your act. That will *download* the file (not show it in-browser).

### CODE REVIEW

**Comment by longhotsummer** [comment]
Resolves #3

**Comment by longhotsummer** [comment]
Requires https://github.com/laws-africa/indigo/pull/425 for Indigo

**Comment by longhotsummer** [comment]
Sample XML files are in the team drive, under Countries > Kenya > Samples > Raw XML. The Accountants act is a reasonable one to base the coverpage XML off of.

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<coverpage>{{ document.title }}</coverpage>
```
Comment (longhotsummer) [inline indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html]: XML goes here.

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html [hunk 2]
```diff
@@ -1 +1,85 @@
...
+    <p style="center">ARRANGEMENT OF SECTIONS</p>
+    <container id="AOS_01" name="AOS_01">
+      {% for item in aos %}
+        <p>
+        {{ item.type }} {{ item.num }} {{ item.heading }}
+        </p>
+        {% if item.children %}
+          {% for child in item.children %}
+            <p>
+            {{ child.type }} {{ child.num }} {{ child.heading }}
```
Comment (longhotsummer) [inline indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html]: i wonder if `{{ child.title }}` would be useful here. It's the formatted title that the TOC builder comes up with for the element. It does nice things like `2. Section Title` for sections and `Part 2 - Foo` for parts.

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html [hunk 3]
```diff
@@ -1 +1,85 @@
...
+    </container>
+  </container>
+  <p style="center">
+    <b>NO. {{ document.work.number }} OF {{ document.work.year }}</b>
+  </p>
+  <p>
+    <docTitle>{{ document.title|upper }}</docTitle>
+  </p>
+  <p style="right">
+    [
```
Comment (longhotsummer) [inline indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html]: The newlines will introduce spaces between the `[` and the text. You probably want to pull those onto the same line. When viewing XML in a browser that makes it look pretty, it puts all tags onto a new line.

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html (hunk 2 above)
Comment (goose-life) [reply indigo_kenyalaw/templates/indigo_kenyalaw/akn/coverpage-legacy.xml.html]: `.title` and `.heading` are giving me the same thing

File: indigo_kenyalaw/renderers.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        act.meta.addnext(coverpage)
+
+    def generate_coverpage(self, document):
+        xml = render_to_string("indigo_kenyalaw/akn/coverpage-legacy.xml.html", {
+            'document': document,
+        })
+        return xml
+
+    def update_preface(self, document, act):
+        old_preface_content = act._get('act.preface').p
```
Comment (longhotsummer) [inline indigo_kenyalaw/renderers.py]: It's not going to be quite this easy. First, be sure to handle when there is no preface. Second, we need to serialize the XML into a string, and then strip the `<preface...>` and `</preface>` that it'll put in there. You can then pass it into your template as you're doing.

Use this to serialise:

``etree.tostring(act.act.preface)``

File: indigo_kenyalaw/renderers.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        })
+        return xml
+
+    def flat_aos(self, aos):
+        flat_aos = []
+        item_id = 1
+        blocklist_id = 1
+        list_introduced = False
+
+        for item in aos:
```
Comment (longhotsummer) [inline indigo_kenyalaw/renderers.py]:
```python
def flatten(entry):
    flat_aos.append(entry)
    if entry.children:
        for child in entry.children:
            flatten(child)
```

File: indigo_kenyalaw/renderers.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                            flat_aos.append(subpart)
+                            if subpart.children:
+                                for cross_head in subpart.children:
+                                    flat_aos.append(cross_head)
+                                    if cross_head.children:
+                                        for section in cross_head.children:
+                                            flat_aos.append(section)
+
+        for i, item in enumerate(flat_aos):
+            if item.type == 'section':
```
Comment (longhotsummer) [inline indigo_kenyalaw/renderers.py]: not even going to try to wrap my head around this on friday at 3pm!

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/preface-legacy.xml.html [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  <p style="right">
+    [<i>Date of assent:
+      <docDate date="{{ document.work.assent_date|date:'Y-m-d' }}" id="assent">
+        {{ document.work.assent_date|date:" dS F, Y" }}</docDate></i>.]
+  </p>
+  <p style="right">
+    [<i></i>]
+  </p>
+  <longTitle>
+    <p>{{ old_preface_content }}</p>
```
Comment (longhotsummer) [inline indigo_kenyalaw/templates/indigo_kenyalaw/akn/preface-legacy.xml.html]: If you apply the above change, then `old_preface_content` can simply be put in just after the `</p>` above.

**Comment by longhotsummer** [comment]
Nice, this is looking good!

File: indigo_kenyalaw/templates/indigo_kenyalaw/akn/preface-legacy.xml.html (hunk 7 above)
Comment (longhotsummer) [reply indigo_kenyalaw/templates/indigo_kenyalaw/akn/preface-legacy.xml.html]: We'll obviously have to provide better support for longTitle.

File: indigo_kenyalaw/renderers.py (hunk 5 above)
Comment (goose-life) [reply indigo_kenyalaw/renderers.py]: oooooh thank you!

Comment (longhotsummer) [reply indigo_kenyalaw/renderers.py]: Let's put recursion on that skills matrix :)

**Pull request merged**